### DISCUSSION TITLE
TransmissionData complex data structures

### DISCUSSION THREAD

**Discussion by antaradey-itk** [eclipse-iceoryx/iceoryx2#159] [Q&A] [answered]
I am trying to send a hashmap data to the subscriber, which looks like this:

pub struct TransmissionMapData {
    pub data : HashMap<String, f64>
}

On the publisher side, the program looks like this: 
fn main() {
         let service = zero_copy::Service::new(&service_name)
        .publish_subscribe()
        .open_or_create::<TransmissionMapData>()?;
   while (tick_event) {
        counter +=1;
        let mut sensor_data = HashMap::new();
        sensor_data.insert("Camera".to_string(), counter as f64);
        let sensor_map_data = TransmissionMapData { data: sensor_data };
        let sample = publisher.loan_uninit()?;
        let sample: iceoryx2::sample_mut::SampleMut<'_, '_, '_, zero_copy::Service<'_>, 
        iceoryx2::service::header::publish_subscribe::Header, TransmissionMapData> =
        sample.write_payload(TransmissionMapData { data: sensor_data });
        println!("Send sample {} ...", counter);
        println!("Sent: {:?} ...", sample.payload());
        publisher.send(sample)?;
}
..
}

On the subscriber side, 
fn main() {
         let service = zero_copy::Service::new(&service_name)
        .publish_subscribe()
        .open_or_create::<TransmissionMapData>()?;
    while (tick_event) {
        while let Some(sample) = subscriber.receive()? {
            println!("received: {:?}", *sample); //or println!("received: {:?}", sample.payload());
        }
}
..
}

The issue is: The subscriber does not fetch any value, and the program gets corrupted meaning SIGKILL needs to be sent to kill the process.
On the publisher side, the send sample works fine, and payload prints correctly.

Is there anything specific I might be missing here? I also tried serialising/Deserialising to Json_string, but looks like it isn't necessary, as the MessageType can handle other data structures, though the subscriber process does not work. Please provide insights into this.

**Comment by elBoberido** [comment, marked as answer]
There are restriction on the data which can be used in shared-memory for zero-copy inter-process communication. I'm not sure whether we already documented it for iceoryx2 but it is basically the same as for iceoryx1 https://github.com/eclipse-iceoryx/iceoryx/blob/master/doc/website/getting-started/overview.md#restrictions. Furthermore the structs should be `repr(C)` to prevent issues due to reordering of fields.

We plan to create a trait similar to `Send` and a derive macro which checks whether a struct contains only fields which can be used in shared-memory. It will probably be done for the v0.3 or v0.4 release.

**Reply by antaradey-itk** [reply]
Okay. 
As it does not allow heap memory, that is why I was unable to serialize using serde, as by default, it serializes into a heap Json String. 
So, I used postcard crate to serialize the map data structure into heapless::vec, and now it works! :)
Thanks for the help!
